drivers/adc/adc_mcp356x.c

```diff
@@ -0,0 +1,836 @@
+{
+	struct mcp356x_data *data = dev->data;
+	enum adc_action action = ADC_ACTION_CONTINUE;
+
+	if (data->ctx.sequence.options && data->ctx.options.callback) {
+		action = data->ctx.options.callback(dev, &data->ctx.sequence,
+						    data->ctx.sampling_index);
+	}
+	if (action == ADC_ACTION_CONTINUE) {
+		if (!data->ctx.sequence.options ||
+		    ++data->ctx.sampling_index > data->ctx.options.extra_samplings) {
+			action = ADC_ACTION_FINISH;
+		}
+	}
+	if (action == ADC_ACTION_FINISH) {
+		fast_write(dev, CMD_STANDBY);
+		adc_context_complete(&data->ctx, 0);
+		return false;
+	}
+	adc_context_update_buffer_pointer(&data->ctx, action == ADC_ACTION_REPEAT);
+	return true;
+}
+
+#else /* CONFIG_ADC_MCP356X_USE_INTERNAL_TIMER */
+
+/* In system timer mode, each sample is one-shot.  We just call adc_context_on_sampling_done(),
+ * and it starts the next sample if necessary.
+ */
+static bool next_samples(const struct device *dev)
+{
+	struct mcp356x_data *data = dev->data;
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+	return false; /* No more samples, it's always one at a time */
+}
+#endif /* CONFIG_ADC_MCP356X_USE_INTERNAL_TIMER */
+
+/* Thread waits for adc_context_start_sampling(), then does one conversion sequence when kicked.
+ * In continuous mode, loops until the sequence finishes.  In one-shot mode, does one sample,
+ * then goes back to waiting for adc_context_start_sampling().
+ *
+ * Sampling is done by sending the start command, waiting for the drdy irq, and then reading the
+ * data out, which is in sample_channels().
+ */
+static void mcp356x_acquisition_thread(const struct device *dev)
+{
+	struct mcp356x_data *data = dev->data;
+
+	while (true) {
+		k_sem_take(&data->acq_sem, K_FOREVER);
+		/* Begin a new seqeunce */
+
+		uint32_t channels = data->ctx.sequence.channels;
+		const int bps = data->ctx.sequence.resolution / 8;
+
+		fast_write(dev, CMD_START);
+
+		/* In continuous mode, this will loop until finished with the sequence.  In
+		 * one-shot mode this will run once as next_samples() always returns false.i
+		 */
+		int ret = 0;
+
+		do {
+			ret = sample_channels(dev, channels, bps);
+		} while (!ret && next_samples(dev));
+
+		if (ret) {
+			/* Failure, we need to stop the sequence and the ADC */
+			if (IS_ENABLED(CONFIG_ADC_MCP356X_USE_INTERNAL_TIMER))
+				fast_write(dev, CMD_STANDBY);
+			adc_context_complete(&data->ctx, ret);
+		}
+	}
+}
+
+static int mcp356x_channel_setup(const struct device *dev,
+				 const struct adc_channel_cfg *cc)
+{
+	struct mcp356x_data *data = dev->data;
+	const struct mcp356x_config *config = dev->config;
+
+	/* Verify some things before we start programming the device registers */
+
+	if (cc->channel_id > 16 || cc->reference != ADC_REF_EXTERNAL0) {
+		return -EINVAL;
+	}
+	if (cc->input_positive >= 16 || cc->input_negative >= 16) {
+		return -EINVAL;
+	}
+	if (!cc->differential) {
+		LOG_WRN("The MCP356x only supports differential mode");
+		return -EINVAL;
+	}
+
+	uint8_t gain;
```

Comment:
We generally declare variables at the top of the block (this goes for the entire file).